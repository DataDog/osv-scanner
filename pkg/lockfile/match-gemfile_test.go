package lockfile_test

import (
	"bytes"
	"errors"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"testing"

	"github.com/google/osv-scanner/pkg/lockfile"
	"github.com/google/osv-scanner/pkg/models"
	"github.com/stretchr/testify/assert"
)

var gemfileMatcher = lockfile.GemfileMatcher{}

func TestGemfileMatcher_GetSourceFile_FileDoesNotExist(t *testing.T) {
	t.Parallel()

	lockFile, err := lockfile.OpenLocalDepFile("fixtures/bundler/no-gemfile/Gemfile.lock")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := gemfileMatcher.GetSourceFile(lockFile)
	expectErrIs(t, err, fs.ErrNotExist)
	assert.Equal(t, "", sourceFile.Path())
}

func TestGemfileMatcher_GetSourceFile(t *testing.T) {
	t.Parallel()
	dir, err := os.Getwd()
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	basePath := "fixtures/bundler/one-package/"
	sourcefilePath := filepath.FromSlash(filepath.Join(dir, basePath+"Gemfile"))

	lockFile, err := lockfile.OpenLocalDepFile(basePath + "Gemfile.lock")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := gemfileMatcher.GetSourceFile(lockFile)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	assert.Equal(t, sourcefilePath, sourceFile.Path())
}

func TestGemfileMatcher_Match_OnePackage(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/bundler/one-package/Gemfile")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:    "RedCloth",
			Version: "4.2.9",
		},
	}
	err = gemfileMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	expectPackages(t, packages, []lockfile.PackageDetails{
		{
			Name:    "RedCloth",
			Version: "4.2.9",
			BlockLocation: models.FilePosition{
				Line:     models.Position{Start: 3, End: 5},
				Column:   models.Position{Start: 1, End: 24},
				Filename: sourceFile.Path(),
			},
			NameLocation: &models.FilePosition{
				Line:     models.Position{Start: 3, End: 3},
				Column:   models.Position{Start: 5, End: 15},
				Filename: sourceFile.Path(),
			},
			VersionLocation: &models.FilePosition{
				Line:     models.Position{Start: 4, End: 4},
				Column:   models.Position{Start: 2, End: 12},
				Filename: sourceFile.Path(),
			},
		},
	})
}

func TestGemfileMatcher_OnePackage_MatcherFailed(t *testing.T) {
	dir, err := os.Getwd()
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	stderr := os.Stderr
	r, w, err := os.Pipe()
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}
	os.Stderr = w

	matcherError := errors.New("gemfileMatcher failed")
	lockfile.GemfileExtractor.Matcher = FailingMatcher{Error: matcherError}

	path := filepath.FromSlash(filepath.Join(dir, "fixtures/bundler/one-package/Gemfile.lock"))
	packages, err := lockfile.ParseGemfileLock(path)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	// Capture stderr
	_ = w.Close()
	os.Stderr = stderr
	var buffer bytes.Buffer
	_, err = io.Copy(&buffer, r)
	if err != nil {
		t.Errorf("failed to copy stderr output: %v", err)
	}
	_ = r.Close()

	assert.Contains(t, buffer.String(), matcherError.Error())
	expectPackagesWithoutLocations(t, packages, []lockfile.PackageDetails{
		{
			Name:      "RedCloth",
			Version:   "4.2.9",
			Ecosystem: lockfile.BundlerEcosystem,
			CompareAs: lockfile.BundlerEcosystem,
		},
	})

	// Reset buildGradleMatcher mock
	MockAllMatchers()
}
