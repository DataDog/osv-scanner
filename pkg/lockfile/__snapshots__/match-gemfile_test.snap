
[TestGemfileMatcher_Match_Groups - 1]
[
  {
    "name": "thor",
    "version": "1.3.2",
    "blockLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 1,
        "end": 20
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "versionLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 13,
        "end": 20
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "nameLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 5,
        "end": 11
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "timeout",
    "version": "0.4.3",
    "depGroups": [
      "production"
    ],
    "blockLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 1,
        "end": 76
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "versionLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 16,
        "end": 23
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "nameLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 5,
        "end": 14
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "useragent",
    "version": "0.16.11",
    "depGroups": [
      "default",
      "production"
    ],
    "blockLocation": {
      "line": {
        "start": 10,
        "end": 11
      },
      "column": {
        "start": 1,
        "end": 15
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "versionLocation": {
      "line": {
        "start": 10,
        "end": 10
      },
      "column": {
        "start": 18,
        "end": 27
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "nameLocation": {
      "line": {
        "start": 10,
        "end": 10
      },
      "column": {
        "start": 5,
        "end": 16
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "websocket-driver",
    "version": "",
    "depGroups": [
      "development"
    ],
    "blockLocation": {
      "line": {
        "start": 16,
        "end": 16
      },
      "column": {
        "start": 3,
        "end": 42
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "nameLocation": {
      "line": {
        "start": 16,
        "end": 16
      },
      "column": {
        "start": 7,
        "end": 25
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "websocket-extensions",
    "version": "0.1.5",
    "depGroups": [
      "development",
      "test"
    ],
    "blockLocation": {
      "line": {
        "start": 23,
        "end": 23
      },
      "column": {
        "start": 3,
        "end": 38
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "versionLocation": {
      "line": {
        "start": 23,
        "end": 23
      },
      "column": {
        "start": 31,
        "end": 38
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "nameLocation": {
      "line": {
        "start": 23,
        "end": 23
      },
      "column": {
        "start": 7,
        "end": 29
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "zeitwerk",
    "version": "2.7.1",
    "depGroups": [
      "development",
      "test"
    ],
    "blockLocation": {
      "line": {
        "start": 24,
        "end": 24
      },
      "column": {
        "start": 3,
        "end": 43
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "versionLocation": {
      "line": {
        "start": 24,
        "end": 24
      },
      "column": {
        "start": 19,
        "end": 26
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "nameLocation": {
      "line": {
        "start": 24,
        "end": 24
      },
      "column": {
        "start": 7,
        "end": 17
      },
      "file_name": "/Users/rodrigo.fernandes/go/src/github.com/DataDog/osv-scanner/pkg/lockfile/fixtures/bundler/groups/Gemfile"
    },
    "packageManager": "Bundler"
  }
]
---
